### PR TITLE
feat(repository): add Git::Repository.open/.bare factories and path state

### PR DESCRIPTION
**Pull request by jcouball**
# Description

Implements architectural redesign **Step C1a-1** as defined in
`redesign/3_architecture_implementation.md`.

This is a **4.x-compatible, additive** step. It introduces new factory scaffolding
on `Git::Repository` and relocates path-resolution logic out of `Git::Base` into a
dedicated module, **without changing any top-level entry-point behavior**:
`Git.open` and `Git.bare` continue to return `Git::Base`.

## What changed

- **New factories:** `Git::Repository.open(working_dir, options = {})` and
  `Git::Repository.bare(git_dir, options = {})` construct a `Git::Repository`
  backed by a `Git::ExecutionContext::Repository` built from resolved paths.
- **New module `Git::Repository::PathResolver`** (`lib/git/repository/path_resolver.rb`):
  owns `resolve_paths` and the public `root_of_worktree`, plus private helpers
  (`resolve_working_directory`, `resolve_repository`, `resolve_gitdir_pointer`,
  `resolve_index`, `execute_rev_parse_toplevel`). This logic was moved from
  `Git::Base`, with one deliberate correction: `resolve_gitdir_pointer` now
  resolves relative `gitdir:` targets against the directory containing the pointer
  file (`File.dirname(path)`), matching Git's own specification. The prior
  `Git::Base` implementation expanded relative targets against `working_dir`, which
  was incorrect. All real `gitdir:` pointer files written by `git` use absolute
  paths in the scenario where the pointer file lives outside the working-tree root,
  so this correction has no practical effect on existing callers.
- **New accessors on `Git::Repository`:** `#dir`, `#repo`, `#index` (each a
  `Pathname` or `nil`, derived from the execution context as the single source of
  truth) and `#repo_size` (Integer byte total).
- **`Git::Base` now delegates:** `.open`/`.bare`/`.clone` call
  `Git::Repository::PathResolver.resolve_paths`, and the public
  `Git::Base.root_of_worktree` class method delegates to
  `Git::Repository::PathResolver.root_of_worktree`. The now-duplicated private
  helpers were removed from `Git::Base`, and a dead
  `require 'git/commands/rev_parse'` (made unused by this move) was removed.

## Why this change is needed

This is foundational scaffolding for the larger Strangler Fig migration of
`Git::Base`/`Git::Lib` into the `Git::Repository` facade. Establishing the path
state and `.open`/`.bare` factories on `Git::Repository` first — while keeping
`Git::Base` entry-point behavior identical for all practical callers — lets
subsequent steps move public methods over incrementally with no user-visible
disruption.

## Backward compatibility

Audited with the **review-backward-compatibility** skill — **verdict: backward
compatible**:

- `Git::Base.root_of_worktree` keeps its signature, return value, and raising
  behavior (legacy `tests/units/test_git_base_root_of_worktree.rb` passes).
- `.open`/`.bare`/`.clone` resolve `@working_directory`/`@repository`/`@index`
  identically for all practical inputs; only the call receiver changed. The one
  functional difference is that `resolve_gitdir_pointer` now correctly resolves
  relative `gitdir:` targets against the pointer file's own directory rather than
  `working_dir`, fixing a latent bug. Because `git` itself always writes absolute
  paths in this situation, no real-world caller is affected.
- The removed private methods were `private_class_method` and had no external
  references.
- A pre-existing `Git::Base#repo_size` bug (`Pathname#path` is protected) predates
  this step and is intentionally left out of scope; the new
  `Git::Repository#repo_size` implements the equivalent algorithm correctly via
  `repo.to_s`.

## Test coverage

- **Unit tests:** `spec/unit/git/repository_spec.rb`,
  `spec/unit/git/repository/path_resolver_spec.rb` — factories, all four
  accessors (incl. `nil`/bare branches), and `resolve_paths`/`root_of_worktree`
  edge cases (gitdir pointer absolute/relative/plain-file, bare defaults,
  `Errno::ENOENT`, `Git::FailedError`, nonexistent dir).
- **Integration tests:** `spec/integration/git/repository_spec.rb` — real `.open`/
  `.bare` construction against real git, accessor values, and a regression guard
  that `Git.open`/`Git.bare` still return `Git::Base`.
- **Coverage:** 100% line **and** branch coverage of both
  `lib/git/repository.rb` and `lib/git/repository/path_resolver.rb` (verified via
  merged SimpleCov resultset).

## Quality verification

- `bundle exec rake default:parallel` (CI-equivalent: tests + RuboCop + YARD +
  build) passes with no failures, no offenses, and no Ruby warnings.
- Full RSpec (6021 examples) and Test::Unit (556 tests) suites green.
- YARD: 0 warnings, coverage 83.6% (threshold 75%); all new public API documented.

## Breaking changes

None. This step is additive and 4.x-compatible.

## Checklist

- [x] I reviewed and applied the project's [AI Policy](../AI_POLICY.md). I understand
  and verify any AI-assisted changes included in this PR and ensured they meet
  quality, security, and licensing standards.
- [x] I ran `bundle exec rake` locally on this branch and it passed (see
  [Local development setup](../CONTRIBUTING.md#local-development-setup)).
- [x] Tests added/updated as needed.
- [x] Documentation updated where applicable (README and/or YARD).

### CODE REVIEW

File: lib/git/repository/path_resolver.rb [hunk 1]
```diff
@@ -0,0 +1,194 @@
+# frozen_string_literal: true
+
+require 'git/commands/rev_parse'
+
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/repository/path_resolver.rb:6]: Removed reference to Git::Lib.

File: lib/git/repository/path_resolver.rb [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    # `PathResolver` is the single home for the path-resolution logic used by the
+    # `Git::Repository` factory class methods ({Git::Repository.open} and
+    # {Git::Repository.bare}). It computes the absolute working-directory,
+    # repository (`.git`), and index paths from the caller-supplied values,
+    # following the same rules Git itself uses (including gitdir-pointer files for
+    # submodules and linked worktrees).
+    #
+    # All methods are pure with respect to their inputs except
+    # {.root_of_worktree}, which shells out to `git rev-parse --show-toplevel`.
+    #
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/repository/path_resolver.rb]: Removed this statement.

File: lib/git/repository/path_resolver.rb [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+      #
+      def resolve_gitdir_pointer(path)
+        return path unless File.file?(path)
+
+        gitdir_content = File.read(path).strip
+        return path unless gitdir_content.start_with?('gitdir: ')
+
+        gitdir_path = gitdir_content.sub(/\Agitdir: /, '')
+        File.expand_path(gitdir_path, File.dirname(path))
+      end
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/repository/path_resolver.rb:187]: The Git::Base method was a bug. This new implementation fixes it. The observation is correct — this is a deliberate behavior change, and the PR description's "moved verbatim" language is inaccurate on this specific point. Apologies for the misleading wording.

 I'll update the PR description to remove the "moved verbatim / behaviorally identical" claim and replace it with a note that this one behavior was corrected in the move.

File: lib/git/repository.rb [hunk 4]
```diff
@@ -55,6 +58,127 @@ class Repository
     include Git::Repository::StatusOperations
     include Git::Repository::WorktreeOperations
 
+    # Open a working copy or bare repository at an existing path
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/repository.rb]: Yes, fix it.

File: lib/git/repository.rb [hunk 5]
```diff
@@ -55,6 +58,127 @@ class Repository
...
+    #   auto-detected).
+    #
+    # @option options [String] :index a non-standard path to the index file
+    #
+    # @option options [Logger] :log a logger forwarded to the command layer
+    #
+    # @option options [String, nil] :git_ssh path to a custom SSH executable
+    #
+    # @option options [String] :binary_path path to the git binary
+    #
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/repository.rb]: Yes, this is correct.

File: lib/git/repository.rb [hunk 6]
```diff
@@ -55,6 +58,127 @@ class Repository
...
+    # @param git_dir [String] the path to the bare repository directory
+    #
+    # @param options [Hash] options forwarded to the constructed repository
+    #
+    # @option options [Logger] :log a logger forwarded to the command layer
+    #
+    # @option options [String, nil] :git_ssh path to a custom SSH executable
+    #
+    # @option options [String] :binary_path path to the git binary
+    #
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/repository.rb]: Yes, fix this.

**Comment by jcouball** [comment]
Thanks for the review! Addressed all four points in `a887dd93`:

1. **Double `lstat` / single-pass / `Errno::ENOENT` (`repo_size`)** — Rewrote `repo_size` to traverse once with `Find.find`, taking a single `File.lstat` per entry and accumulating `stat.size` directly (no intermediate `contained_files` array, no second `lstat`). Added a per-entry `rescue Errno::ENOENT` so files that vanish mid-scan are skipped instead of failing.

2. **Symlinked-directory traversal counting external files** — `Find.find` uses `lstat` and does **not** descend into symlinked directories, so files that physically live outside the repository can no longer be reached/counted. This replaces the previous `Dir.glob('**/*')` + string-prefix containment approach entirely (the fragile `repo_prefix`/`expand_path` guard is gone).

3. **Misleading `does not exist` message (`root_of_worktree`)** — `Dir.exist?` is also false for a regular file, so the message now reads `'<path>' does not exist or is not a directory` (and the `@raise` YARD tag was updated to match). The legacy `/does not exist/` assertions still hold.

4. **Missing symlinked-directory regression test** — Added a unit example that creates a symlinked directory inside the repo pointing at an external dir and asserts its files are not counted. Also added: a "file disappears during traversal" example (covers the `ENOENT` rescue) and a `root_of_worktree` example for a path that is a file rather than a directory. Removed the now-obsolete `Dir.glob`-stub `..`-escape test (it exercised the old implementation).

Both `lib/git/repository.rb` and `lib/git/repository/path_resolver.rb` remain at 100% line and branch coverage; RuboCop, YARD, and the full RSpec + Test::Unit suites are green.
